src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -1076,19 +1074,24 @@ MulticopterPositionControl::update_smooth_takeoff(const float &z_sp, const float
 		}
 
 		// Ramp up takeoff speed.
-		_takeoff_speed += desired_tko_speed * _dt / _takeoff_ramp_time.get();
-		_takeoff_speed = math::min(_takeoff_speed, desired_tko_speed);
+		if (_takeoff_ramp_time.get() > _dt) {
+			_takeoff_speed += desired_tko_speed * _dt / _takeoff_ramp_time.get();
+
+		} else {
+			// No ramp, directly go to desired takeoff speed
+			_takeoff_speed = desired_tko_speed;
+		}
+
+		_takeoff_speed = math::min(_takeoff_speed, _tko_speed.get());
 
 		// Smooth takeoff is achieved once desired altitude/velocity setpoint is reached.
 		if (!_smooth_velocity_takeoff) {
 			_in_smooth_takeoff = _states.position(2) - 0.2f > math::max(z_sp, _takeoff_reference_z - MPC_LAND_ALT2.get());
 
-		} else  {
-			_in_smooth_takeoff = _takeoff_speed < -vz_sp;
+		} else {
+			// Make sure to stay in smooth takeoff if takeoff has not been detected yet by the land detector
+			_in_smooth_takeoff = (_takeoff_speed < -vz_sp) || _vehicle_land_detected.landed;
```

Comment:
It doesn't make sense to stay in smooth takeoff when the desired velocity setpoint after the initial step was already reached. It doesn't matter if the vehicle is still on the ground or not. This will result in a downwards step of the velocity as soon as the ramp is over because the takeoff was detected by the land detector.

I'm assuming you inserted this because you lowered the threshold to start the takeoff to a value so low (`_smooth_velocity_takeoff = PX4_ISFINITE(vz_sp) && vz_sp < -0.1f;` from above) that the vehicle stays on the ground and correctly still reports landed.